vpr/src/base/partition.h

```diff
@@ -0,0 +1,52 @@
+#ifndef PARTITION_H
+#define PARTITION_H
+
+#include "vtr_strong_id.h"
+#include "region.h"
+#include "partition_regions.h"
+
+/**
+ * @file
+ * @brief This file defines the data for a partition: a grouping of atoms that are constrained to a portion of an FPGA. A partition defines
+ * the atoms that are assigned to it, and the locations in which they can be placed. The locations are a union of rectangles in x, y,
+ * sub_tile space. Common cases include a single rectangle or a single x, y, sub_tile (specific location). More complex partitions
+ * with L, T or other shapes can be created with a union of multiple rectangles.
+ */
+
+//Type tag for PartitionId
+struct partition_id_tag;
+
+//A unique identifier for a partition
+typedef vtr::StrongId<partition_id_tag> PartitionId;
+
+class Partition {
+  public:
+    PartitionId get_partition_id();
+    void set_partition_id(PartitionId _part_id);
+
+    std::string get_name();
+    void set_name(std::string _part_name);
+
+    //add an id to the atom_blocks vector
+    void add_to_atoms(AtomBlockId atom_id);
+
+    //check if a given atom is in the partition
+    bool contains_atom(AtomBlockId atom_id);
+
+    //get the atom_blocks vector
+    std::vector<AtomBlockId> get_atoms();
+
+    //set the union of regions for this partition;
+    void set_part_regions(PartitionRegions pr);
```

Comment:
This is a nit, but I think it is worth consider.  In general, plural methods like `add_atoms` or `set_part_regions` imply an array context.  However in this case, the array context is a scalar object "PartionRegions".  Does it make more sense to rename `PartitionRegions` to `PartitionCollection`, which is a non-plural name, and makes clear it is a scalar that contains a collection?

Reply:
To illustrate what I'm talking about, consider `atom_blocks`, which is plural and is an array-like object.  However part_regions is also plural, but is a scalar object.

Reply:
Could rename PartitionRegions as PartitionRegion, and PartitionRegion as RectPlusSubtile (or something like that).  Now PartitionRegion is singular, indicating it stores a single (possibly complex) region, and RectPlusSubtile indicates the basic elements used to describe a (possibly composite) region.

Reply:
Right, Region i.e. RectPlusSubtile does not store a composite region - it stores a single rectangular area and optionally a subtile. So changing the name would specify this.

I do think it would be better to rename PartitionRegions to something non-plural since it is a collection of regions belonging to one particular partition. PartitionCollection could work, I could comment that the class stores the collection of regions (RectPlusSubtile's) belonging to the partition.

Reply:
I think PartitionCollection is also the wrong name.  This is really storing the (possibly complex) region that goes with one partition.  So it really should be called PartitionRegion.  PartitionRegions and PartitionCollection both let part of the implementation (currently a union of simple rectangles) leak through the name, which isn't great.  If we someday change to more complex region representations like those Keith suggests, PartitionRegion still works as a name, but PartitionCollection doesn't work as well.